spirv_glsl.cpp

```diff
@@ -3258,6 +3344,107 @@ void CompilerGLSL::emit_resources()
+	if (!options.vulkan_semantics && declared_KHR_shader_subgroup_ballot)
+	{
+		statement("#ifdef GL_NV_shader_thread_group");
+		statement("uvec4 subgroupBallot(bool v) { return uvec4(ballotThreadNV(v), 0u, 0u, 0u); }");
+		statement("#elif defined(GL_ARB_shader_ballot)");
+		statement("uvec4 subgroupBallot(bool v) { return uvec4(unpackUint2x32(ballotARB(v)), 0u, 0u); }");
+		statement("#endif");
+
+
+		statement("#ifdef GL_NV_shader_thread_shuffle");
+
+		for (const char *t : TYPES)
+			statement(string(t) + " subgroupBroadcastFirst(" + t +
+			          " value) { return shuffleNV(value, 0, gl_WarpSizeNV); }");
+		for (const char *t : TYPES)
+			statement(string(t) + " subgroupBroadcast(" + t +
+			          " value, uint id) { return shuffleNV(value, id, gl_WarpSizeNV); }");
+
+		statement("#elif defined(GL_ARB_shader_ballot)");
+
+		for (const char *t : TYPES)
+			statement(string(t) + " subgroupBroadcastFirst(" + t + " value) { return readInvocationARB(value); }");
+		for (const char *t : TYPES)
+			statement(string(t) + " subgroupBroadcast(" + t +
+			          " value, uint id) { return readInvocationARB(value, id); }");
+
+		statement("#endif");
+
+
+		statement("#ifdef GL_NV_shader_thread_group");
+
+		statement("uint subgroupBallotFindLSB(uvec4 value) { return findLSB(value.x); }");
+		statement("uint subgroupBallotFindMSB(uvec4 value) { return findMSB(value.x); }");
+
+		statement("#else");
+
+		statement("uint subgroupBallotFindLSB(uvec4 value) {");
+		statement("int firstLive = findLSB(value.x);");
+		statement("return uint(firstLive != -1 ? firstLive : (findLSB(value.y) + 32));");
+		statement("}");
+		statement("uint subgroupBallotFindMSB(uvec4 value) {");
+		statement("int firstLive = findMSB(value.y);");
+		statement("return uint(firstLive != -1 ? (firstLive + 32) : findMSB(value.x));");
+		statement("}");
+
+		statement("#endif");
+	}
+	// GL_KHR_shader_subgroup_vote
+	if (!options.vulkan_semantics && declared_KHR_shader_subgroup_vote)
+	{
+		statement("#ifdef GL_ARB_shader_group_vote");
+		statement("bool subgroupAll(bool v) { return allInvocationsARB(v); }");
+		statement("bool subgroupAny(bool v) { return anyInvocationARB(v); }");
+		statement("bool subgroupAllEqual(bool v) { return allInvocationsEqualARB(v); }");
+		statement("#define SUBGROUP_ALL_EQUAL_WORKAROUND(type) bool subgroupAllEqual(type value) { return "
+		          "subgroupAllEqual(subgroupBroadcastFirst(value)==value); }");
+		for (const char *t : TYPES)
+		{
+			statement(std::string("SUBGROUP_ALL_EQUAL_WORKAROUND(") + t + ")");
+		}
+		statement("#undef SUBGROUP_ALL_EQUAL_WORKAROUND");
+
+		statement("#elif defined(GL_NV_gpu_shader5)");
+		statement("bool subgroupAll(bool value) { return allThreadsNV(value); }");
+		statement("bool subgroupAny(bool value) { return anyThreadNV(value); }");
+		statement("bool subgroupAllEqual(bool value) { return allThreadsEqualNV(value); }");
+
+		statement("#elif defined(GL_AMD_gcn_shader)&&defined(GL_AMD_gpu_shader_int64)");
```

Comment:
`gcn_shader` requires ARB extension 64-bit integers, not AMD one, IIRC

Reply:
https://www.khronos.org/registry/OpenGL/extensions/AMD/AMD_gcn_shader.txt

> AMD_gpu_shader_int64 or NV_gpu_shader5 is required.

Reply:
ah crap, thanks.